fsm.go

```diff
@@ -308,40 +313,65 @@ func (f *FSM) Event(event string, args ...interface{}) error {
 		return UnknownEventError{event}
 	}
 
-	e := &Event{f, event, f.current, dst, nil, args, false, false}
-
-	err := f.beforeEventCallbacks(e)
-	if err != nil {
+	e := (&Event{
+		FSM:      f,
+		Event:    event,
+		Src:      f.current,
+		Dst:      dst,
+		Err:      nil,
+		Args:     args,
+		canceled: false,
+		async:    false,
+	}).Clone(true)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	e.cancelFunc = cancel
+
+	if err := f.beforeEventCallbacks(ctx, e); err != nil {
 		return err
 	}
 
 	if f.current == dst {
-		f.afterEventCallbacks(e)
+		f.afterEventCallbacks(ctx, e)
 		return NoTransitionError{e.Err}
 	}
 
 	// Setup the transition, call it later.
-	f.transition = func() {
-		f.stateMu.Lock()
-		f.current = dst
-		f.stateMu.Unlock()
-
-		f.enterStateCallbacks(e)
-		f.afterEventCallbacks(e)
+	transitionFunc := func(async bool) func() {
```

Comment:
We'll probably have to add `ctx context.Context` to this will break with async state transitions (context used currently probably timed out or was canceled). There might be a test case we can add – see https://github.com/looplab/fsm/pull/77#issuecomment-930173282.

Reply:
This is done.